### ISSUE TITLE
Unable to open file in RHIDE (EOVERFLOW error)

### ISSUE THREAD

**Issue by MaxPower2005** [stsp/dosemu2#933]
Hello
I'm a long time user of dosemu 1.4.x and RHIDE (yes, I know it is old, but it works for my purposes).
Under dosemu2 (I tried both the devel and master branch), I'm able to start RHIDE without a problem, but as soon as I try to open a file, RHIDE throws a EOVERFLOW error.
I have never observed such a behaviour under the old dosemu. 
Any ideas  what's going wrong?

Steps to reproduce the behaviour:
- Install RHIDE (ftp://ftp.delorie.com/pub/djgpp/current/v2apps/rhid15ab.zip)
- Start RHIDE
- Press F3 to open a file, select file, -> ![RHIDE EOVERFLOW](https://user-images.githubusercontent.com/30479037/66384957-f16d2c80-e9bf-11e9-9eca-cac0f17d9e67.png)

[boot.log](https://github.com/stsp/dosemu2/files/3701693/boot.log)

**Comment by stsp** [comment]
![Снимок экрана от 2019-10-08 14-46-03](https://user-images.githubusercontent.com/3295627/66393604-aa8d3000-e9db-11e9-998d-d7da5237c4a1.png)

Can't reproduce this.
This comment:
https://github.com/stsp/dosemu2/issues/572#issuecomment-388991121
mentions some FPU state fix, but I would
need to reproducer.
I tried to even open some large binary files -
all works.

**Comment by stsp** [comment]
Please attach the `-D9+M` log.

**Comment by stsp** [comment]
There are 2 things to try.

Use `./default-configure -d` to avoid `-O2`
during build. Some additional flags to disable
SSE may help either.

Check if your drive C is read-only by any chance
(as I think it is). In that case `TEMP` var may
point to r/o space, which confuses rhide.

**Comment by MaxPower2005** [comment]
Hi
thanks for your quick response.

> Please attach the -D9+M log.

Sure you really want me to do that. The file is ~6MiB in size.

> Use ./default-configure -d to avoid -O2 during build.

RHIDE still throws EOVERFLOW.

> Some additional flags to disable SSE may help either.

Are you talking about the -mno-sse option of GCC or something else?

> Check if your drive C is read-only by any chance.

I can write to the TEMP directory from the DOS prompt without problem.

**Comment by stsp** [comment]
> Sure you really want me to do that. The file is ~6MiB in size.

Even gzipped?

> Are you talking about the -mno-sse option of GCC or something else?

Mainly -march=i486. Can also use -mno-sse just to be sure.

> I can write to the TEMP directory from the DOS prompt without problem.

You are right and the log was wrong.
I fixed the minor bug that produced wrong logging.

**Comment by MaxPower2005** [comment]
> Even gzipped?

[boot.log.rhide.gz](https://github.com/stsp/dosemu2/files/3706319/boot.log.rhide.gz)

> Mainly -march=i486. Can also use -mno-sse just to be sure.

Still EOVERFLOW in RHIDE.

**Comment by stsp** [comment]
Oh I've noticed you use 32bit build.
Can you check if this is 32bit-specific?

**Comment by stsp** [comment]
Also could you please check it on a few git
tags, like 2.0pre8-dev, 2.0pre7-dev, 2.0pre6.1-dev?

**Comment by MaxPower2005** [comment]
> Also could you please check it on a few git tags, like 2.0pre8-dev, 2.0pre7-dev, 2.0pre6.1-dev?

2.0pre8-dev works.
After a longer "git bisect" session, I narrowed the problem down to the following two commits:

https://github.com/stsp/dosemu2/commit/63b1d77d92e5682208789e6065c1f73a169795ab
https://github.com/stsp/dosemu2/commit/67c4b9a3e18a769a5839c241c0212451ddcc9aeb

Two different commits because with the first one RHIDE does not start at all, therefore I can't tell if the issue is present or not.

**Comment by stsp** [comment]
Thanks, this is very helpful!
Lets try the try-and-error approach.
I created the branch "try" and reverted one
of those patches there.
Could you please give this "try" a try?

**Comment by stsp** [comment]
If the branch "try" works, then it was likely
broken multiple times. Please then try commit
6bbe54c00b which pre-dates both the above
ones. If that doesn't work too, please bisect
between it and pre8-dev.

**Comment by MaxPower2005** [comment]
A quick check shows that your "try" branch "works"; at least the EOVERFLOW error is gone.
https://github.com/stsp/dosemu2/commit/6bbe54c00b443a676e6df8017aec0e537d100f58 does not.
git bisect between 2.0pre8-dev and https://github.com/stsp/dosemu2/commit/6bbe54c00b443a676e6df8017aec0e537d100f58 identifies https://github.com/stsp/dosemu2/commit/415a28592477a99e024dc9a9338d9baf3a3f5fce as the culprit.

Please take this info with a grain of salt as I had to do an additional change to the code to make RHIDE start again. Perhaps the two symptoms (EOVERFLOW and RHIDE not starting) are somehow related.

In other words:
https://github.com/stsp/dosemu2/commit/b936b31d6302d8cb9adf01a6c461ec55ceb3abc5 -> RHIDE starts and file opening is OK.
https://github.com/stsp/dosemu2/commit/415a28592477a99e024dc9a9338d9baf3a3f5fce -> RHIDE does not start at all.
https://github.com/stsp/dosemu2/commit/415a28592477a99e024dc9a9338d9baf3a3f5fce + change -> RHIDE starts but file opening fails.

**Comment by stsp** [comment]
You mean a change to rhide?
But its not needed on a branch "try", as well
as on master (in the latter case opening fails),
right?

**Comment by stsp** [comment]
Could you please see if `$_set_int_hooks` option
affects the behaviour?

**Comment by stsp** [comment]
Are you loading emufs.sys/emufs.com in config.sys?

**Comment by MaxPower2005** [comment]
> You mean a change to rhide?

Sorry, I should have expressed myself more clearly.
No, RHIDE is unchanged. 
The "change" is to cherry-pick https://github.com/stsp/dosemu2/commit/67c4b9a3e18a769a5839c241c0212451ddcc9aeb

> But its not needed on a branch "try", as well as on master (in the latter case opening fails), right?

Yes, "try" does not need this "change" to make RHIDE start.
Regarding master I have to revoke my previous statement. master is working fine (no RHIDE start problem, no EOVERFLOW).

> Are you loading emufs.sys/emufs.com in config.sys?

Yes.

> Could you please see if $_set_int_hooks option affects the behaviour?

With $_set_int_hooks = (0) or "emufs.sys"  commented out,, the following message is printed during RHIDE startup:
Load error: no DPMI - Get csdpmi*b.zip
 
Please note that this is not the same RHIDE startup problem as mentioned before.

**Comment by stsp** [comment]
Master is similar to -pre8 so its of
no interest. Is it possible to give me
an ssh access to analyze the
problem on devel? Its good that we
found the fix, but the fix is quite large
and I dont know why it is still needed.

**Comment by stsp** [comment]
I patched a suspicious place of the aforementioned
patch, please see if anything changed.

**Comment by MaxPower2005** [comment]
> I patched a suspicious place of the aforementioned patch, please see if anything changed.

No, still EOVERFLOW.

> Is it possible to give me an ssh access to analyze the problem on devel?

Over the weekend, I tried to setup two Debian based VMs (32bit and 64bit)  to faciliate debugging of this issue and maybe grant remote access.
Unfortunately, this opened a can of worms full of new problems. Not sure yet if some or all of them are related to dosemu2.

**Comment by stsp** [comment]
Ok, in this case in msdos_xtra() comment
out all 3 occurences of
if (!(old_flags &CF))
NOCARRY;

And see what changes.
This should give some more info.

**Comment by MaxPower2005** [comment]
If I comment out the code as instructed, the EOVERFLOW is gone.
And it seems to be enough to comment out the first occurence only to make RHIDE happy.

**Comment by stsp** [comment]
Thanks, this is a good hint.
But still not enough for me to understand the problem.
Please try the below for some more info:
```
--- a/src/base/core/int.c
+++ b/src/base/core/int.c
@@ -1683,8 +1683,10 @@ static void msdos_xtra(uint16_t old_ax, uint16_t old_flags)
            /* mfs_lfn() clears CF on success, sets on failure, preserves
             * on unsupported */
            ret = mfs_lfn();
-           if (!ret)
+           if (!ret) {
                LWORD(eax) = 0x7100;
+               CARRY;
+           }
        }
        break;
     case 0x73:
```

Does it change anything?

**Comment by MaxPower2005** [comment]
With this change, the EOVERFLOW is gone too.

**Comment by stsp** [comment]
How interesting!
In this same "if" clause please print
the value of "old_eax" and "old_flags"
variables and post the results here.
For printing you can use function
error() - its like printf in dosemu.

**Comment by MaxPower2005** [comment]
> the value of "old_eax" and "old_flags"

I assume you mean "old_ax", not "old_eax", correct?

Below is the requested output:
ERROR: old_ax: 0x71A6, old_flags: 0x3000
ERROR: old_ax: 0x71A6, old_flags: 0x3000
ERROR: old_ax: 0x71A6, old_flags: 0x3000
ERROR: old_ax: 0x71A6, old_flags: 0x3000

**Comment by stsp** [comment]
Thanks, should now be fixed.
Please check.

**Comment by MaxPower2005** [comment]
Sorry but it still does not work.
I still have to add a "CARRY" in the first "if (!ret) { }" in msdos_xtra() to make it work.

**Comment by stsp** [comment]
Oh, impossible! :)
Can there be a second bug in the same a6 func? :(
Please try this and see is something changes:
```
--- a/src/dosext/mfs/mfs.c
+++ b/src/dosext/mfs/mfs.c
@@ -3116,7 +3116,7 @@ void get_volume_label(char *fname, char *fext, char *lfn, int drive)
 char *sft_to_filename(unsigned char *sft, int *fd)
 {
   int cnt = sft_fd(sft);
-  *fd = open_files[cnt].name ? open_files[cnt].fd : 0;
+  *fd = open_files[cnt].name ? open_files[cnt].fd : HANDLE_INVALID;
   return open_files[cnt].name;
 }
 
```

**Comment by MaxPower2005** [comment]
No change.

**Comment by stsp** [comment]
Please see if adding CARRY the way you
did the last time, still fixes the problem even
in presence of the aforementioned change.
From now on, please check that CARRY
thing in the case of "no change" otherwise.

Another thing to check (w/o CARRY and with CARRY):
```
--- a/src/dosext/mfs/lfn.c
+++ b/src/dosext/mfs/lfn.c
@@ -1168,7 +1168,7 @@ static int mfs_lfn_(void)
                d_printf("LFN: get file info by handle %x\n", _BX);
                filename = handle_to_filename(_BX, &fd);
                if (filename == NULL)
-                       return fd ? lfn_error(fd) : 0;
+                       return lfn_error(HANDLE_INVALID);
 
                if (fstat(fd, &st))
                        return lfn_error(HANDLE_INVALID);
```

Thanks!

**Comment by MaxPower2005** [comment]
> Please see if adding CARRY the way you did the last time, still fixes the problem even
in presence of the aforementioned change

Sorry I forgot to mention previously it but yes, adding the "CARRY" still fixes the issue.

> Another thing to check

Shall I add this on top of a clean https://github.com/stsp/dosemu2/commit/8126f43f0180d4c3cc9990b00d07ae6039651754 or to my already modified tree?

**Comment by stsp** [comment]
Both, to clean and to CARRY mod.

**Comment by MaxPower2005** [comment]
With your last change on top of https://github.com/stsp/dosemu2/commit/8126f43f0180d4c3cc9990b00d07ae6039651754 RHIDE is happy (no "CARRY" needed).

**Comment by stsp** [comment]
Please test this one (on clean branch):
```
--- a/src/dosext/mfs/lfn.c
+++ b/src/dosext/mfs/lfn.c
@@ -105,10 +105,12 @@ static char *handle_to_filename(int handle, int *fd)
                return NULL;
 
        /* do we "own" the drive? */
-       *fd = 0;
        dd = sft_device_info(sft) & 0x0d1f;
-       if (dd < 0 || dd >= MAX_DRIVE || !drives[dd].root)
+       if (dd < 0 || dd >= MAX_DRIVE || !drives[dd].root) {
+               error("dd=%i\n", dd);
                return NULL;
+       }
+       *fd = 0;
 
        return sft_to_filename(sft, fd);
 }
```

**Comment by MaxPower2005** [comment]
With this change RHIDE is happy too (no "CARRY" needed).

Output:
ERROR: dd=19
ERROR: dd=19
ERROR: dd=19
ERROR: dd=19

**Comment by stsp** [comment]
Please attach `-D9+dD#` log.

**Comment by stsp** [comment]
From 4eb322434c

**Comment by stsp** [comment]
Please test the `try2` branch.
I just removed that code there, so if still no
change, then I don't know what. :)

**Comment by stsp** [comment]
That will likely not work though.
I'll try to rewrite this code later.
The problem is that here rhide doesn't
execute that code (but for you it does),
so I can't properly test.

**Comment by stsp** [comment]
OK, branch `try2` is waiting for a test.

**Comment by MaxPower2005** [comment]
RHIDE is working ok (no EOVERFLOW) with https://github.com/stsp/dosemu2/commit/3be5394b93b34e7640064d4df8688a9f278e2f50

Do you need a `-D9+dD#` log of this test?

**Comment by stsp** [comment]
Yes, would be nice.
Since I haven't found the bug and just
re-wrote the suspicious code (that I didn't
understand), the log will help me to see that
my new code is at least executed, and therefore
works as expected. For me here it doesn't execute,
so it was written "blindly".

**Comment by MaxPower2005** [comment]
As requested
[boot.log.gz](https://github.com/stsp/dosemu2/files/3733903/boot.log.gz)

**Comment by stsp** [comment]
Thanks!
The log shows that the code is executed and
works as expected.
I'll apply it.
In a mean time I added a few more fixes to devel
(w/o yet applying the aforementioned one) - could
you please test current devel wrt this problem?

**Comment by MaxPower2005** [comment]
Not sure if this is what you have expected but RHIDE is working ok (no EOVERFLOW) with https://github.com/stsp/dosemu2/commit/e33fed7c2fc7b3436cb14223c746fc88ab7b036a

[boot.log.gz](https://github.com/stsp/dosemu2/files/3734431/boot.log.gz)

**Comment by stsp** [comment]
Thanks!
I merged also the remaining fixes now.
